File: AlgorithmsPackCPP/Sequence Reconstruction.cpp
class Solution {
    typedef unordered_map<int,vector<int>>GRAPH;

    void addEdge(GRAPH&graph, int from, int to){
        graph[from].push_back(to);
    }
public:
    vector<int>topSort(GRAPH&graph){
        unordered_map<int,int>incomingEdges;
        for(auto&node:graph){
            for(int j:node.second){
                incomingEdges[j]++;
            }
        }
        queue<int>q;
        for(auto&node:graph){
            if(!incomingEdges.count(node.first))
                q.push(node.first);
        }
        vector<int>result;
        while(!q.empty()){
            if(q.size()>1)
                return vector<int>();

            int cur = q.front();
            q.pop();
            result.push_back(cur);

            for(int j:graph[cur]){
                if(--incomingEdges[j]==0){
                    q.push(j);
                }
            }
        }
        if(result.size()!=graph.size())
            result.clear();
        return result;
    }

    bool sequenceReconstruction(vector<int>& nums, vector<vector<int>>& sequences) {
        GRAPH graph(nums.size());
        for(auto&seq:sequences){
            for(int i = 0; i < seq.size()-1; i++){
                addEdge(graph,seq[i],seq[i+1]);
            }
        }
        for(auto&seq:sequences){
            if(seq.size()==1 && graph.count(seq[0])==0){
                graph[seq[0]] = vector<int>();
            }
        }
        return topSort(graph)==nums;
    }
};

File: AlgorithmsPackPythonSolutions/Sequence Reconstruction.py
class Solution:
    def add_edge(self, graph, from_node, to):
        graph[from_node].append(to)

    def topSort(self, graph):
        indegrees = defaultdict(int)

        for node in graph.values():
            for j in node:
                indegrees[j] += 1

        q = deque()
        for node in graph:
            if node not in indegrees:
                q.append(node)

        ordering = []

        while q:
            if len(q) > 1:
                return []
            cur = q.popleft()
            ordering.append(cur)

            for j in graph[cur]:
                indegrees[j] -= 1
                if indegrees[j] == 0:
                    q.append(j)

        if len(ordering) != len(graph):
            return []
        return ordering

    def sequenceReconstruction(self, nums: List[int], sequences: List[List[int]]) -> bool:
        graph = defaultdict(list)

        for seq in sequences:
            for i in range(len(seq) - 1):
                self.add_edge(graph, seq[i], seq[i + 1])

        for seq in sequences:
            if len(seq) == 1 and seq[0] not in graph:
                graph[seq[0]] = []

        ordering = self.topSort(graph)
        return ordering == nums
